transports/tls-quic/src/certificate.rs

```diff
@@ -275,6 +269,157 @@ impl P2pCertificate<'_> {
         // if signature verification fails.
         self.extension.public_key.verify(&msg, &self.extension.signature)
     }
+
+    /// Return the signature scheme corresponding to [`AlgorithmIdentifier`]s
+    /// of `subject_pki` and `signature_algorithm`
+    /// according to https://tools.ietf.org/id/draft-ietf-tls-tls13-21.html#rfc.section.4.2.3.
+    pub fn signature_scheme(&self) -> Option<rustls::SignatureScheme> {
+        // Certificates MUST use the NamedCurve encoding for elliptic curve parameters.
+        // Endpoints MUST abort the connection attempt if it is not used.
+        use rustls::SignatureScheme::*;
+        use oid_registry::*;
+        let signature_algorithm = &self.certificate.signature_algorithm;
+        let pki_algorithm = &self.certificate.tbs_certificate.subject_pki.algorithm;
+
+        if pki_algorithm.algorithm == OID_PKCS1_RSAENCRYPTION {
+            if signature_algorithm.algorithm == OID_PKCS1_SHA256WITHRSA {
+                return Some(RSA_PKCS1_SHA256)
+            }
+            if signature_algorithm.algorithm == OID_PKCS1_SHA384WITHRSA {
+                return Some(RSA_PKCS1_SHA384)
+            }
+            if signature_algorithm.algorithm == OID_PKCS1_SHA512WITHRSA {
+                return Some(RSA_PKCS1_SHA512)
+            }
+            if signature_algorithm.algorithm == OID_PKCS1_RSASSAPSS {
+                // FIXME move OIDs to `oid_registry`
+                #[allow(non_snake_case)]
+                let OID_NIST_HASH_SHA384 = der_parser::oid::Oid::from(&[2,16,840,1,101,3,4,2,2])
+                    .expect("This is a valid OID; qed");
+                #[allow(non_snake_case)]
+                let OID_NIST_HASH_SHA512 = der_parser::oid::Oid::from(&[2,16,840,1,101,3,4,2,3])
+                    .expect("This is a valid OID; qed");
+
+                // According to https://datatracker.ietf.org/doc/html/rfc4055#section-3.1:
+                // Inside of params there shuld be a sequence of:
+                // - Hash Algorithm
+                // - Mask Algorithm
+                // - Salt Length
+                // - Trailer Field
+
+                // We are interested in Hash Algorithm only, however the der parser parses
+                // params into a mess, so here is a workaround to fix it:
+                let params = signature_algorithm.parameters.as_ref()?;
+                let params = params.as_sequence().ok()?;
+                let first_param = params.get(0)?;
+                let hash_oid_der = first_param.as_slice().ok()?;
+                let (_, obj) = der_parser::parse_der(hash_oid_der).ok()?;
+                let hash_oid = obj.as_sequence().ok()?.get(0)?.as_oid_val().ok()?;
+
+                if hash_oid == OID_NIST_HASH_SHA256 {
```

Comment:
match

Reply:
It wont make the code look better because of https://stackoverflow.com/questions/28225958/why-is-this-match-pattern-unreachable-when-using-non-literal-patterns.